src/DDMinimizer.cpp

```diff
@@ -0,0 +1,445 @@
+// controlled gates
+std::pair<std::map<std::string, std::map<std::pair<Qubit, Qubit>, int>>,
+          std::map<std::string, std::vector<int>>>
+DDMinimizer::makeDataStructure(qc::QuantumComputation& qc) {
+  const std::size_t bits = qc.getNqubits();
+
+  // map string of cX, xC, cL, xL, cR, xR to map of control and target bit
+  // to the index of the gate
+  std::map<std::string, std::map<std::pair<Qubit, Qubit>, int>> maps;
+
+  // string of cX, xC, cL, xL, cR, xR to vector of max index of gate
+  // for cL and xL position 0 in the vector marks the line of c(x) at 0 -> we
+  // count the left most as the first for cR and xR position 0 in the vector
+  // marks the line of c(x) at bits - 1 -> we count the right most as the first
+  std::map<std::string, std::vector<int>> indices;
+
+  // create x-c ladder
+  const std::size_t max = bits - 1;
+
+  // maps for the x-c and c-x ladder with pair of control qubit and target qubit
+  // to index
+  std::map<pair<Qubit, Qubit>, int> xCMap;
+  std::map<pair<Qubit, Qubit>, int> cXMap;
+  for (size_t i = 0; i < max; i++) {
+    xCMap.insert({{i + 1, i}, -1});
+    cXMap.insert({{i, i + 1}, -1});
+  }
+  // save complete maps of ladder and for max index evaluation
+  maps.insert({"xC", xCMap});
+  indices.insert({"xC", std::vector<int>(1, 0)});
+  maps.insert({"cX", cXMap});
+  indices.insert({"cX", std::vector<int>(1, 0)});
+
+  // create c-l and x-l ladder
+  for (size_t i = 0; i < max; i++) {
+    std::map<pair<Qubit, Qubit>, int> xLMap;
+    std::map<pair<Qubit, Qubit>, int> cLMap;
+    // create the steps of the ladder
+    for (size_t j = 0; j < bits; j++) {
+      if (i < j) {
+        xLMap.insert({{j, i}, -1});
+        cLMap.insert({{i, j}, -1});
+      }
+    }
+    // save the steps in the maps
+    maps.insert({"xL" + std::to_string(i), xLMap});
+    maps.insert({"cL" + std::to_string(i), cLMap});
+  }
+  // save the complete ladder for max index evaluation
+  indices.insert({"xL", std::vector<int>(bits - 1, 0)});
+  indices.insert({"cL", std::vector<int>(bits - 1, 0)});
+
+  // create c-r and x-r ladder
+  for (size_t i = max; i > 0; i--) {
+    std::map<pair<Qubit, Qubit>, int> xRMap;
+    std::map<pair<Qubit, Qubit>, int> cRMap;
+    // create the steps of the ladder
+    for (size_t j = 0; j < bits; j++) {
+      if (i > j) {
+        xRMap.insert({{j, i}, -1});
+        cRMap.insert({{i, j}, -1});
+      }
+    }
+    // save the steps in the maps
+    maps.insert({"xR" + std::to_string(i), xRMap});
+    maps.insert({"cR" + std::to_string(i), cRMap});
+  }
+  // save the complete ladder for max index evaluation
+  indices.insert({"xR", std::vector<int>(bits - 1, 0)});
+  indices.insert({"cR", std::vector<int>(bits - 1, 0)});
+
+  return std::make_pair(maps, indices);
+}
+
+// Helper function to check if the vector of a ladder step is full, meaning each
+// gate appeared in the circuit
+bool DDMinimizer::isFull(const std::vector<int>& vec) {
```

Comment:
Could you give this function a more expressive name? Maybe something like `isFullLadder`?